### PR TITLE
Feat: add heuristics to miner for nakamoto

### PR DESCRIPTION
**Pull request by kantai**
This add two heuristics for mining in nakamoto:

* For the first block in a tenure, just mine an empty block
* Estimate the time it takes to eval a tx, and see if it will interfere with block deadline. If it will, skip the transaction.

Marking as draft until tests can be fleshed out and likely broken integration tests fixed.

### CODE REVIEW

**Review by jcnelson — COMMENTED** [review]
I think the idea is sound, but testing this in a reproducible way is gonna be interesting.

File: stackslib/src/core/mempool.rs [hunk 1]
```diff
@@ -2650,6 +2659,24 @@ impl MemPoolDB {
         Ok(())
     }
 
+    /// Drop and blacklist transactions, so we don't re-broadcast them or re-fetch them.
+    /// Do *NOT* remove them from the bloom filter.  This will cause them to continue to be
+    /// reported as present, which is exactly what we want because we don't want these transactions
+    /// to be seen again (so we don't want anyone accidentally "helpfully" pushing them to us, nor
+    /// do we want the mempool sync logic to "helpfully" re-discover and re-download them).
```
Comment (obycode) [inline stackslib/src/core/mempool.rs:2666]: I think this is accidentally copied from the method below.

**Review by obycode — COMMENTED** [review]
LGTM! I like this addition! Just noticed one rogue comment.

File: stackslib/src/chainstate/stacks/miner.rs [hunk 2]
```diff
@@ -2250,10 +2261,27 @@ impl StacksBlockBuilder {
...
                         }
-                        if get_epoch_time_ms() >= deadline {
+                        let time_now = get_epoch_time_ms();
+                        if time_now >= deadline {
                             debug!("Miner mining time exceeded ({} ms)", max_miner_time_ms);
                             return Ok(None);
                         }
+                        if let Some(time_estimate) = txinfo.metadata.time_estimate_ms {
+                            if time_now.saturating_add(time_estimate.into()) > deadline {
+                                debug!("Mining tx would cause us to exceed our deadline, skipping";
```
Comment (obycode) [inline stackslib/src/chainstate/stacks/miner.rs:2271]: It might be nice to have this as an `info!`.

**Review by obycode — DISMISSED** [review]
👍

File: stackslib/src/core/mempool.rs [hunk 3]
```diff
@@ -803,6 +803,16 @@ const MEMPOOL_SCHEMA_6_NONCES: &'static [&'static str] = &[
     "#,
 ];
 
+const MEMPOOL_SCHEMA_7_TIME_ESTIMATES: &'static [&'static str] = &[
+    r#"
+    -- ALLOW NULL
+    ALTER TABLE mempool ADD COLUMN time_estimate_ms NUMBER;
```
Comment (jcnelson) [inline stackslib/src/core/mempool.rs:809]: Wouldn't the column affinity be `INTEGER`?

File: stackslib/src/core/mempool.rs [hunk 4]
```diff
@@ -803,6 +803,16 @@ const MEMPOOL_SCHEMA_6_NONCES: &'static [&'static str] = &[
...
+    "#,
+    r#"
+    INSERT INTO schema_version (version) VALUES (7)
```
Comment (jcnelson) [inline stackslib/src/core/mempool.rs:812]: Should be `UPDATE schema_version SET version = 7` so we don't get a bunch of rows.

Comment (jcnelson) [reply stackslib/src/core/mempool.rs:812]: Oh I see -- this is already a problem :/

**Comment by kantai** [comment]
> The changes look fine to me, but I'm curious if the mempool changes are sufficient to stave off a network stall in fuzzing. The code will only detect that a transaction would exceed its designated processing time after it has been run once. But if the fuzzer is bombarding the network with lots and lots of expensive transactions, which are only meant to ever be evaluated once, then this wouldn't materially change anything.

That's mostly true, but its also an argument against incremental improvement. Does this change address all possible long running transactions in the mempool? No, but it does at least prevent the same long running transaction from repeatedly preventing the miner from making progress.

> Do we know for sure that the cause of block production slowness in fuzzing is due to re-considering the same transaction over and over?

In the particular case that this addresses, yes.

> Also, would this change mean that some transactions are never mined, despite being valid, simply because they were considered once and effectively blocked from re-consideration due to taking too long to run? When it comes time to rolling this out into production, we'll want to be sure that this doesn't lead to some contracts being effectively blocked from executing.

That would be ultimately up to the miner and signer sets. Having an estimate for how long a transaction will take to run and comparing against the miners self-configured execution deadline doesn't seem like a policy change vis-a-vis long-running transactions: rather, it just makes the miner somewhat smarter about hitting its deadling.

**Comment by jcnelson** [comment]
> That's mostly true, but its also an argument against incremental improvement. Does this change address all possible long running transactions in the mempool? No, but it does at least prevent the same long running transaction from repeatedly preventing the miner from making progress.

Apologies if I sounded like I was against incremental improvement.  I was only trying to confirm my own understanding.

File: stackslib/src/core/mempool.rs (hunk 3 above)
Comment (kantai) [reply stackslib/src/core/mempool.rs]: Fixed in [6f60813](https://github.com/stacks-network/stacks-core/pull/5238/commits/6f60813c91468dce9971a4a859c98f86dc31a1e0)

**Comment by kantai** [comment]
New tests and assertions added for the behaviors, and all existing tests pass. Ready for review!

**Review by obycode — APPROVED** [review]
Lgtm!

**Review by jcnelson — APPROVED** [review]

**Pull request merged**